File: dynamic_programming/longest_ideal_string.cs
public class Solution {
    public int LongestIdealString(string s, int k) {
        const char a = 'a';
        const int maxChars = 25;

        Span<int> dp = stackalloc int[maxChars + 1];
        int max = 1;

        for (int i = 0; i < s.Length; ++i) {
            int temp = 0;
            for (int j = Math.Max(0, s[i] - a - k); j <= Math.Min(maxChars, s[i] - a + k); j++) {
                temp = Math.Max(temp, dp[j]);
            }

            dp[s[i] - a] = temp + 1;
            max = Math.Max(max, temp + 1);
        }

        return max;
    }
}

File: dynamic_programming/longest_ideal_string.cpp
class Solution {
public:
    int longestIdealString(string s, int k) {
        const char a = 'a';
        const int max_chars = 25;

        vector<int> dp(max_chars + 1, 0);

        for (const char& c : s) {
            int tmp = 0;
            int x = c - a;
            for (int i = max(0, x - k); i <= min(max_chars, x + k); ++i) {
                tmp = max(tmp, dp[i]);
            }

            dp[x] = ++tmp;
        }

        return *max_element(begin(dp), end(dp));
    }
};

static auto init = []() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  cout.tie(nullptr);
  return 'c';
}();
